src/vos/vos_query.c

```diff
@@ -203,25 +198,58 @@ query_recx(struct open_query *query, daos_recx_t *recx)
 	if (rc != 0)
 		goto fini;
 
-	recx->rx_idx = entry.en_sel_ext.ex_lo;
-	recx->rx_nr = entry.en_sel_ext.ex_hi - entry.en_sel_ext.ex_lo + 1;
+	if (opc & EVT_ITER_SKIP_HOLES) {
+		if (entry.en_visibility & EVT_VISIBLE) {
+			recx->rx_idx = entry.en_sel_ext.ex_lo;
+			recx->rx_nr = entry.en_sel_ext.ex_hi - entry.en_sel_ext.ex_lo + 1;
+		} else {
+			rc = -DER_NONEXIST;
+		}
+		D_GOTO(fini, rc);
+	}
+
+	/* The following only supports MAX recx query at the moment */
+	if (!(query->qt_flags & VOS_GET_MAX))
+		D_GOTO(fini, rc);
+
+	/* let's find the last punched extent */
+	while (!bio_addr_is_hole(&entry.en_addr)) {
+		rc = evt_iter_next(ih);
+		if (rc)
+			goto fini;
+		rc = evt_iter_fetch(ih, &inob, &entry, NULL);
+		if (rc != 0)
+			goto fini;
+	}
+
+	/* The last punched extent */
+	end = entry.en_sel_ext.ex_hi;
+	while (bio_addr_is_hole(&entry.en_addr)) {
+		if (start == 0 || entry.en_sel_ext.ex_hi == start - 1)
+			start = entry.en_sel_ext.ex_lo;
+		else
+			break;
+		rc = evt_iter_next(ih);
+		if (rc)
```

Comment:
may check fail case of "rc != -DER_NONEXIST".
And not sure if VOS has a quick check that if the hole existed or not first, then only need to do the while() loop for hole exists case.

Reply:
not sure what you mean? in any case, if Iter_next() return failure, it means it either it reach to the end or it met failure, and we break for both cases.

Reply:
in your previous version, seems the rc is not checked (two while()), fine for this version.